tools/building.py

```diff
@@ -350,6 +350,22 @@ def lld_flags_for_executable(external_symbols):
   return cmd
 
 
+def no_whole_archive_for_rust_libs(args):
+  ret = []
```

Comment:
It makes me a little sad to see such a rust-specific hack here, but I can see how it solves your issue.

Can I ask about your use case?   Do you know why it is that you need to use `MAIN_MODULE=1`?   Would it be possible to use `MAIN_MODULE=2` and then explicitly list all the symbols you want exported in a file?

If you really do need `MAIN_MODULE=1` when do you not also need all the symbols from the `.rlib` libraries to be exported too?   Normally `MAIN_MODULE=1` means that the set of symbol that might be needs cannot be known and therefore everything gets exported... but what about the symbols in the `.rlib` files?

If you want `MAIN_MODULE=1`, but only for certain libraries then perhaps you could try injecting the `--whole-archive` flags yourself on top of `MAIN_MODULE=2` build.

Something like this:

```
$ ./emcc -sMAIN_MODULE=1 libfoo.rlib -Wl,--whole-archive libineedallofthis.a -Wl,--no-whole-archive -O2 -g `-Wl,--whole-archive
```

The final whole-archive should be used if you want to include all the system library code

Reply:
That way you would be explict about which libraries you want to force include rather than implying that based on the fact that a library ends with the `.rlib` extension.

Reply:
Maybe we should be using `-sSIDE_MODULE=2` in a lot of places. I should look into that. `-sMAIN_MODULE=2` definitely can't work for us but I didn't realize until recently that we could potentially use `-sMAIN_MODULE=1` with `-sSIDE_MODULE=2`.

Reply:
Regarding `-sSIDE_MODULE=2`, yes that would be a great solution.

With both `-sSIDE_MODULE=2` and `-sMAIN_MODULE=2` you can still include that entire contents of a given library if you want (with `--whole-archive` around just that library).. it just gives you more control.

Reply:
I think rust already generates `-sEXPORTED_FUNCTIONS=[whatever functions]` as part of the linker invocation.

Reply:
Well we have a significant amount of trouble generating controlling the linker invocation because they are almost always generated by other tools. We use a compiler wrapper to do complicated argument rewriting before calling emcc but in the long term it's good to reduce the complexity of the compiler wrapper. But I think `-sSIDE_MODULE=2` will work without any other changes.

Reply:
Thanks!